### PR TITLE
chore(common/models, web): more script updates

### PR DESCRIPTION
**Pull request by jahorton**
This updates the build and test scripts for both Web's `input-processor` and the lm-layer to a builder-based format, which involved some significant tweaking here and there.  

Also fixes an issue I discovered with the error-trap handler that appeared when trying to handle node-based errors; simply re-emitting the error seems a simple enough resolution there.  Without the trap's deliberate re-emission of the exit code, this was occurring during test runs in certain failure states:

image

(I discovered this when a `require` statement in one of the test suites failed.)

One of the more interesting problems I had to solve here was how to handle the chained user-testing setup we've been maintaining here for the conveniences of the related CI test-builds.  I ended up settling on the idea of adding a third target: `:libraries`.  A generalized `./test.sh test` will test all targets - `:libraries`, `:headless`, and `:browser`.  However, when developing, `./test.sh test:headless` and `./test.sh test:browser` will both bypass library-focused tests, allowing a bit of a developer time-save and shortcut.

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
Hmm... now comes the question of how to tweak our CI scripts while updating pre-existing scripts...

https://build.palaso.org/viewLog.html?buildId=341733&buildTypeId=Keyman_Common_LMLayer_TestPullRequests&tab=buildLog

```
Step 1/3: Build full LMLayer module (Command Line) (2s)
[13:07:16]
[Step 1/3] Starting: C:\BuildAgent\temp\agentTmp\custom_script1083849156662489547.cmd
[13:07:16]
[Step 1/3] in directory: C:\BuildAgent\work\99b311828f4ee7c\keyman\common\predictive-text
[13:07:19]
[Step 1/3] build.sh: invalid parameter: -clean
```

Well, now that `clean` is an action, not an option, something will need to change here.

**Comment by mcdurdin** [comment]
See the Keyman Core test builds. We test for the presence of a .builder file in the same folder. Writing on a phone from a plane so hard to look just now.

**Comment by jahorton** [comment]
> See the Keyman Core test builds. We test for the presence of a .builder file in the same folder. Writing on a phone from a plane so hard to look just now.

Self note:  findable under the `epic-ldml` branch @ https://github.com/keymanapp/keyman/blob/7baa9f01cd8813ff2014bb59c4f0034f631633d3/core/.build-builder.

File: common/predictive-text/build.sh [hunk 1]
```diff
@@ -20,92 +20,147 @@ THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BA
...
+################################ Main script ################################
+
+builder_check_color "$@"
+
+builder_describe "Builds the lm-layer module" \
+  "clean" \
+  "configure" \
+  "build" \
+  "test" \
+  ":libraries  Targets all in-repo libraries that this module is dependent upon" \
```
Comment (mcdurdin) [inline common/predictive-text/build.sh:32]: Conceptually, `configure` is supposed to configure and build any internal and extenal dependencies, so then you don't need the `:libraries` target.

The primary issue here is: that `npm i`/`npm ci` is just so painfully slow and noisy, so we really can't afford to run it for each dependency, which would keep their `configure` action clean and isolated. If we can find a good clean method of only running `npm ci` _once_ per command-line invocation of `build.sh configure` (that is, if we called `verify_npm_setup` more than once, it would only run `npm ci` once), I think that really does cover most of our bases. This, I think, does it (and removes need for `fetch_deps` parameter to `verify_npm_setup`):

```bash
# Installs npm dependencies, but only once per build invocation
verify_npm_setup () {
  if [ -z ${_builder_verify_npm_setup_run+x} ]; then
    export _builder_verify_npm_setup_run=true

    # Check if Node.JS/npm is installed.
    type npm >/dev/null ||\
      fail "Build environment setup error detected!  Please ensure Node.js is installed!"

    init_npm
  fi
}
```

There is also overlap between `common/web/keyman-version/build.sh` and `tsc -b` locally, because `tsc -b` builds all dependencies if they are out of date. But as you know, keyman-version has a required pre-build step which `tsc -b` does not support. So we have to run the build script.

I am formulating a better approach which in theory would be passed into `builder_describe`. But have not yet completed that design...

File: common/predictive-text/build.sh [hunk 2]
```diff
@@ -20,92 +20,147 @@ THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BA
...
-echo "Typescript compilation successful."
+if builder_start_action clean :browser; then
+  if [ -n "$CLEANED" ]; then
+    echo "${BUILDER_TERM_START}clean${BUILDER_TERM_END} already completed as ${BUILDER_TERM_START}${CLEANED}${BUILDER_TERM_END}; skipping."
+  else
+    do_clean
+    CLEANED=clean:browser
+  fi
+  builder_finish_action success clean :browser
+fi
```
Comment (mcdurdin) [inline common/predictive-text/build.sh:105]: It's slightly problematic that `clean` of one target also cleans the other target. Is there any way we can keep them isolated?

File: common/predictive-text/unit_tests/test.sh [hunk 3]
```diff
@@ -9,114 +9,168 @@ THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BA
...
-  echo "  -CI                    to perform continuous-integration friendly tests and reporting"
-  echo "  -headless              to disable the in-browser tests"
-  echo "  -integrated            to disable the 'headless' test suite"
-  echo "  -skip-package-install  (or -S) skips dependency updates"
-  echo "  -? | -h | -help        to display this help information"
-  echo ""
-  exit 0
-}
+# This script runs from its own folder
+cd "$SCRIPT_ROOT"
```
Comment (mcdurdin) [inline common/predictive-text/unit_tests/test.sh:13]:
```suggestion
cd "$THIS_SCRIPT_PATH"
```

Given this is defined in build-utils.sh, let's converge on that name.

File: resources/builder.inc.sh [hunk 4]
```diff
@@ -188,6 +188,9 @@ _builder_failure_trap() {
     fi
 
     builder_finish_action failure $action $target
+
+    # Ensure that the exit code propagates fully.
+    exit $trappedExitCode
   fi
```
Comment (mcdurdin) [inline resources/builder.inc.sh:194]: Can we pull this fix into a separate PR?

**Review by mcdurdin — COMMENTED** [review]
Started review, more to come

File: common/predictive-text/.build-builder [hunk 5]
```diff
@@ -0,0 +1,4 @@
+The presence of this file tells CI to use the new builder_ style parameters for build.sh.
+
+Once all branches for 16.0+ are updated to merge the chore/web/more-script-updates branch, then we
+can remove this file and the corresponding bash test branches in CI.
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/predictive-text/.build-builder:4]: Ah, so we're not (currently) planning to make a separate config like with `/core/`?  The version you referenced in an earlier comment contained text along these lines, so I just made the text more appropriate to this location.

File: common/predictive-text/build.sh (hunk 1 above)
Comment (jahorton) [reply common/predictive-text/build.sh:32]: Part of the reason for the `:libraries` target is because of how it provides a wonderful dev-feature when unit testing the various web projects.  In particular, I personally want the `test:libraries` action to exist; isolating aspects of `test`-only build parts to a `build:libraries` is a happy add-on.

If I've been working on a bug in the main `/web/source` code, I could care less what test results I get from `keyboard_processor` and `input_processor` unit tests - they're completely irrelevant.  Much of the time, I'm also not looking to recompile any test-dependency code, as with Web's recorder modules/tools.  They're often "build once" as far as many parts of the codebase are concerned.

As far as `build:libraries` is concerned, one of the benefits is being able to bypass the `npm` step entirely.  Often, I know that there's absolutely no need for even a single `npm ci` call - and one call adds enough delay that I'd rather zero calls than even a single one.  So, with the current structure proposed by this PR's branch, I can skip the `configure` step in order to get zero such calls while still being able to rebuild in-repo parts of the codebase.  (Say, if I made changes to `lm-worker` code and am rebuilding the lm-layer (`common/predictive-text`.)

If the semantics worry you, perhaps something like the following might make sense?
- `configure-env` ("configure environment")
- `configure-libs` ("configure in-repo dependencies")

Though I suppose that would then necessitate an extra option if dropping `:libraries`:
- `--chain`, intended for use with `test`:  also runs in-repo dependency test scripts
    - The nomenclature can be workshopped, of course.

I feel like that proposed option is less elegant, but we could go that route if you prefer.

File: resources/builder.inc.sh (hunk 4 above)
Comment (jahorton) [reply resources/builder.inc.sh:194]: See #7405.

File: common/predictive-text/.build-builder (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/predictive-text/.build-builder]: > Ah, so we're not (currently) planning to make a separate 16.0+ config like appears to be planned for `/core/`?

The only 16.0 config for core is the WASM test build -- because it is new in 16.0. We are trying not to replace existing CI configurations, rather just tweaking them to support new versions as needed. The version number shown in the CI titles is a little confusing, I agree, and we'll probably drop them in future maintenance of CI (once I go ahead and delete 13.0 configs, probably.)

[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/predictive-text/.build-builder]: > The file you referenced in an earlier comment (or, at least the version I found) contained text along these lines, so I just made the text more appropriate to this location.

Yes, that was probably wrong on my part, sorry!

> ... or is it that the same branch that results in the 'outer' if-else would effectively have the same condition as the inner if-else? That... makes a bit more sense, now that I think about it.

And ... you've lost me here!

File: common/predictive-text/build.sh (hunk 1 above)
Comment (mcdurdin) [reply common/predictive-text/build.sh:32]: So, I want targets to be things in the current script. Ideally, a given build.sh will figure out if internal or external dependencies need configuring or building, and only do those if necessary. It would generally never clean or test dependencies.

However, at present, I have not tried to solve the internal dependency question -- just relying on the various dependencies being configured/built as needed by specific temporary steps. And while what you've got here is a good step, I don't think it quite gets to where I want to be.

I will think about this in the next few days because we do need to cross this bridge for dependencies, and I'd like to get this right -- the rest of the builder scripts stuff is reasonably clean so want to try and keep it that way for internal dependencies also.

File: common/predictive-text/build.sh (hunk 2 above)
Comment (mcdurdin) [reply common/predictive-text/build.sh:103]: Yes, separate folders is the correct answer, but agree it's a bit of a yak shave that doesn't need to happen here. Let's look at this as a future maintenance task.

File: resources/builder.inc.sh (hunk 4 above)
Comment (mcdurdin) [reply resources/builder.inc.sh]: Thanks. Will review.

File: common/predictive-text/.build-builder (hunk 5 above)
[the remark this answers is not in the document]
Comment (jahorton) [reply common/predictive-text/.build-builder]: From the CI config:

```
rem Keyman 16 moves Keyman Core to core/
echo Check for core/build.sh [Keyman 16+]
if exist core\build.sh (
  cd core
  if exist .build-builder (
    "C:\Program Files\Git\bin\bash.exe" --init-file "c:\Program Files\Git\etc\profile" -l ./build.sh configure:x86 build:x86 test:x86 configure:x64 build:x64 test:x64
  ) else (
	rem TODO: this branch not required once chore/core/7247-use-builder-for-build-sh changes merged into all branches on 16.0+
    "C:\Program Files\Git\bin\bash.exe" --init-file "c:\Program Files\Git\etc\profile" -l ./build.sh
  )
)
```

Once we've fully merged the referenced changeset into all remaining 16.0+ branches, `if exist core\build.sh` and `if exist .build-builder` will (effectively) be equivalent checks.  So, chalk it up to a "right, for the wrong reasons" re: the contents of the `.build-builder` file - we'll be able to simplify the quoted part of the CI config above.

File: common/predictive-text/build.sh (hunk 1 above)
Comment (jahorton) [reply common/predictive-text/build.sh:32]: I hear you.  I spent longer than I'd prefer to admit before landing at this PR's current "solution" as it was, trying to keep it "reasonably clean."  KMW will want to use a similar approach to whatever we land on here, and I could see similar concerns for the Android & iOS scripts.

Comment (mcdurdin) [reply common/predictive-text/build.sh:32]: > similar concerns

... not to mention Developer which is arguably the worst of all, with dependencies all over the place from web, core, and common.

I'm too tired to be able to do a good job of this tonight, but will continue on this after my vacation.

File: common/predictive-text/.build-builder (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/predictive-text/.build-builder]: For core, yes, we can simplify due to the move from /common/core/desktop/ to /core/, but for other places, we probably can't simplify in the same way -- because build.sh already existed in the same place for 15.0.

**Comment by mcdurdin** [comment]
Web test failures do not appear to be related to these changes but need to be addressed!

image

**Review by mcdurdin — APPROVED** [review]
LGTM together with the builder dep follow-on PRs #7438 and #7439

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.79-alpha](https://keyman.com/downloads/releases/alpha/16.0.79)
